File: Greedy/1282_M_GroupPeopleGivenGroupSizeTheyBelongTo.cxx
/**
 * 5.10 night 5.28 night debug
 * https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/
 */
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    // https://stackoverflow.com/questions/18362896/printing-the-vector-in-a-map
    void print_map(unordered_map<int, vector<int>> map)
    {
        for (auto it = map.begin(); it != map.end(); ++it)
        {
            cout << it->first << " : ";
            for (auto it2 = it->second.begin(); it2 != it->second.end(); ++it2)
                cout << *it2 << " ";
            cout << endl;
        }
    }

    void print(vector<int> const &input)
    {
        for (int i = 0; i < input.size(); i++)
        {
            cout << input.at(i) << ' ';
        }
        cout << endl;
    }

    // https://www.cnblogs.com/qinduanyinghua/p/12010648.html
    vector<vector<int>> groupThePeople_cnblog(vector<int> &groupSizes)
    {

        vector<vector<int>> res;
        unordered_map<int, vector<int>> mp;
        for (int i = 0; i < groupSizes.size(); ++i)
        {
            mp[groupSizes[i]].push_back(i);
        }
        print_map(mp);
        for (auto x : mp)
        {
            int i = x.first;          //i表示组的人数
            vector<int> v = x.second; //可以放在同一个组的所有人的编号
            vector<int> t(i);         //每i个人放一个组
            print(v);
            print(t);
            for (int j = 0; j < v.size(); ++j)
            {
                t[j % i] = v[j];
                if ((j + 1) % i == 0)
                {
                    res.push_back(t);
                }
            }
        }
        return res;
    }

    // https://blog.csdn.net/Eric_1993/article/details/105087187
    vector<vector<int>> groupThePeople_csdn(vector<int> &groupSizes)
    {
        unordered_map<int, vector<int>> group;
        vector<vector<int>> res;
        // 统计拥有相同 group size 的人有哪些
        for (int i = 0; i < groupSizes.size(); ++i)
            group[groupSizes[i]].push_back(i);
        for (auto &p : group)
        {
            auto gs = p.first;
            auto index = p.second;
            // 对于拥有相同 group size (gs) 的这些人, 他们的个数肯定是 gs 的整数倍
            // 那么我们就知道可以将这些人划分为 index.size() / gs 个 Group 了.
            for (int i = 0; i < index.size() / gs; ++i)
            {
                vector<int> tmp;
                std::copy(index.begin() + i * gs, index.begin() + (i + 1) * gs, back_inserter(tmp));
                res.push_back(tmp);
            }
        }
        return res;
    }

    // https://blog.csdn.net/reigns_/article/details/103506291
    vector<vector<int>> groupThePeople_csdn2(vector<int> &groupSizes)
    {
        int n = groupSizes.size();
        if (n == 0)
        {
            return {};
        }
        unordered_map<int, vector<int>> m;
        for (int i = 0; i < n; i++)
        {
            m[groupSizes[i]].push_back(i);
        }
        vector<vector<int>> res;
        for (auto it = m.begin(); it != m.end(); it++)
        {
            int cnt = 0;
            vector<int> tmp;
            for (int j = 0; j < it->second.size(); j++)
            {
                cnt++;
                tmp.push_back(it->second[j]);
                if (cnt == it->first)
                {
                    cnt = 0;
                    res.push_back(tmp);
                    tmp.clear();
                }
            }
        }
        return res;
    }
};

int main()
{
    Solution s;
    vector<int> groupSizes;
    groupSizes.push_back(3);
    groupSizes.push_back(3);
    groupSizes.push_back(3);
    groupSizes.push_back(3);
    groupSizes.push_back(3);
    groupSizes.push_back(1);
    groupSizes.push_back(3);

    s.groupThePeople_cnblog(groupSizes);
    s.groupThePeople_csdn(groupSizes);
    s.groupThePeople_csdn2(groupSizes);
}

File: Greedy/1282_M_GroupPeopleGivenGroupSizeTheyBelongTo.js
/**
 * 5.10 night 5.28 night debug
 * https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/
 */

// https://github.com/mickey0524/leetcode/blob/master/1282.Group-the-People-Given-the-Group-Size-They-Belong-To.java
// Accepted --- 88ms 41.9MB 36.81%
const groupThePeople_mickey0524 = (groupSizes) => {
  let map = new Map();
  let res = [];
  // for (let i = 0; i < groupSizes.length; i++) {
  //   let g = groupSizes[i];
  //   if (!map.has(g)) {
  //     map.set(g, []);
  //   }
  // }
  // console.log(map);  // map no  problem

  for (let i = 0; i < groupSizes.length; i++) {
    let g = groupSizes[i];
    if (!map.has(g)) {
      map.set(g, []);
    }
    // console.log(map);
    let l = map.get(g);
    // console.log("l is: ", l);
    l.push(i);
    if (l.length == g) {
      // console.log("l is: ", l);
      // console.log(typeof l);  

      // res.push(l);
      res.push(Object.values(l)); // fixed problem here, why l is object, should be array

      // console.log("res is: ", res);

      l.splice(0, l.length); // fixed problem here: clear()
      // l = [];
    }
  }
  return res;
};

// https://www.cnblogs.com/qinduanyinghua/p/12010648.html
// not fixed
const groupThePeople_cnblog = (groupSizes) => {
  let res = [];
  let mp = new Map();
  let value = [];
  let uniqueArr = unique(groupSizes);

  for (let i = 0; i < groupSizes.length; ++i) {
    for (const j of uniqueArr) {
      let index = getAllIndexes(groupSizes, j);
      value.push(index);
      // console.log(value);
      mp.set(groupSizes[i], value[i]);
    }
  }
  console.log(mp);
  for (const k of mp.keys()) {
    let i = k;
    let v = mp.get(k);
    let t = i;
    console.log(v); // problem
    console.log(t);
    for (let j = 0; j < v.length; ++j) {
      t[j % i] = v[j];
      if ((j + 1) % i == 0) {
        res.push(t);
      }
    }
  }
  return res;
};

const unique = (arr) => {
  return arr.filter((x, i, a) => {
    return a.indexOf(x) == i
  });
}

const unique2 = (arr) => {
  return [...new Set(arr)];
}

const getAllIndexes = (arr, val) => {
  let indexes = [],
    i = -1;
  while ((i = arr.indexOf(val, i + 1)) != -1) {
    indexes.push(i);
  }
  return indexes;
}

const main = () => {
  let groupSizes = [3, 3, 3, 3, 3, 1, 3];
  let groupSizes2 = [2, 1, 3, 3, 3, 2];
  console.log(groupThePeople_mickey0524(groupSizes));
  console.log(groupThePeople_mickey0524(groupSizes2));

  console.log("");
  console.log(groupThePeople_cnblog(groupSizes));
  // console.log(unique(groupSizes));
  // console.log(unique2(groupSizes));

};

main()